### ISSUE TITLE
[HOLD for payment 2023-06-23] [$2000] Android - Copy icon is missing in Profile/Contact method

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#17368]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open NewDot app
2. Go to profile \ contact method

## Expected Result:
Copy icon should be displayed near [receipts@expensify.com](mailto:receipts@expensify.com)
email

## Actual Result:
Updated copy icon is missing in Profile/Contact method

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.2.99.4

**Reproducible in staging?:** Yes

**Reproducible in production?:** yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/231621215-5a52fd49-070f-47df-9f58-5c862fa3b290.mp4

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0129c789975d0ac1e3</li>
        <li>Upwork Job ID: 1646595570154647552</li>
        <li>Last Price Increase: 2023-05-05</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @lschurr (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by lschurr** [comment]
Confirmed, the copy option isn't there on Android, but it is on iOS.

**Comment by lschurr** [comment]
Should this be internal or external @iwiznia?

**Comment by MelvinBot** [comment]
Triggered auto assignment to @NikkiWines (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by Pujan92** [comment]
I think this is something related to the nested Text element issue in android 
https://github.com/facebook/react-native/issues/31826

**Comment by Balabhadrad** [comment]
SVG

**Comment by MelvinBot** [comment]
📣 @Balabhadrad! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by Balabhadrad** [comment]
Contributor details
Your Expensify account email: balabhadradd@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01c26bfaba14c42f76

**Comment by MelvinBot** [comment]
✅ Contributor details stored successfully. Thank you for contributing to Expensify!

**Comment by MelvinBot** [comment]
@NikkiWines, @lschurr, @aimane-chnaif Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by Victor-Nyagudi** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Android - Copy icon is missing in Profile/Contact method

### What is the root cause of that problem?

I suspect this behavior is unique to Android and is caused by nesting a `View` inside a `Text` component, as is done so here in `CopyTextToClipboard.js`.

https://github.com/Expensify/App/blob/c144663fcad41a69344311a27dd6a147a28e7191/src/components/CopyTextToClipboard.js#L56-L71

The `Tooltip` component has a condition which renders its children whenever the device doesn't support hovering, therefore, the `Icon` component is rendered which wraps the `svg` inside a `View`.

https://github.com/Expensify/App/blob/c144663fcad41a69344311a27dd6a147a28e7191/src/components/Tooltip/index.js#L145-L149

https://github.com/Expensify/App/blob/c144663fcad41a69344311a27dd6a147a28e7191/src/components/Icon/index.js#L54-L65

Interestingly enough, the `Copy` icon isn't missing - it's rendered in the wrong place. 

<details>

<summary>Screenshot of 'copy to clipboard' icon location</summary>

![expensify-copy-to-clipboard-icon-location](https://user-images.githubusercontent.com/79470910/233065798-00d2b977-4fe5-4f3b-98b9-954d2f2687fb.png)

</details>

However, in order to see it, you'll first need to save while in the `CopyTextToClipboard.js` file to trigger a re-render of the component. This leverages [React Native's fast refresh](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&ved=2ahUKEwiF9Pvw8LX-AhVMh_0HHVPqDeUQFnoECAoQAQ&url=https%3A%2F%2Freactnative.dev%2Fdocs%2Ffast-refresh&usg=AOvVaw3XFtBSCWF35NLJFo0AcAgP) feature to re-render just the default-exported component, otherwise, saving in a file higher in the component hierarchy might not re-render the component of interest. 

The icon even responds to click events like it's supposed to whether visible (after it's re-rendered) or invisible (current behavior when app is loaded). 

<details>

<summary>Invisible 'copy to clipboard' icon click interaction</summary>

https://user-images.githubusercontent.com/79470910/233067383-ec1147f0-2c39-40ce-97cb-5746ee394804.mov

</details>

<details>

<summary>Visible 'copy to clipboard' icon click interaction</summary>

https://user-images.githubusercontent.com/79470910/233067541-c9175fa6-b6a7-4d93-a656-385968ed651f.mov

</details>

To demonstrate further, I tried nesting a different `View` inside a `Text` component in `ContactMethodsPage.js` and got similar behavior to the current bug. 

Here's the code I used. 

```javascript
//...
{loginMenuItems}
<View>
    <Text>
       Hello, there 
           <View>
               <Text>I'm nested in a View</Text>
           </View>
    </Text>
</View>
//...
```

<details>

<summary>My result after nesting a View inside a Text component</summary>

![expensify-nested-view-android](https://user-images.githubusercontent.com/79470910/233069647-f53166dd-0278-425b-a52a-295684b5fb5a.png)

</details>

### What changes do you think we should make in order to solve the problem?
This looks like a bug in React Native that may or may not be related to this [currently open React Native Android bug](https://github.com/facebook/react-native/issues/29665) discussing nested `View`s in Android.

I'd suggest not including the 'copy to clipboard' icon on Android devices for the time being, however, the [docs forbid contributors from writing platform-specific code](https://github.com/Expensify/App#philosophy), so I think it would be best to leave implementing this at the discretion of Expensify's internal team. 

Opening an issue with React Native also sounded like a good idea to me.

### What alternative solutions did you explore?

I noticed one of the `View`s wrapping the `svg` was absolutely positioned, so I tried to play around with `left` and `right` style values to see what I could find. 

One of the things I did was add a `backgroundColor` and adjust `height` & `width` of both `View`s wrapping the `svg` and noticed the outermost wrapper's color i.e. `red`, doesn't show on Android but shows on web (presumably on the other platforms where everything is working too).

<details>

<summary>Web result</summary>

![expnsify-copy-icon-with-background](https://user-images.githubusercontent.com/79470910/233073445-cf331840-5b93-443f-935a-5d3be1082c8d.png)

</details>

<details>

<summary>Android result</summary>

![expensify-highlighted-copy-icon-wrapper](https://user-images.githubusercontent.com/79470910/233073563-f52015e2-b41f-4ec7-8a70-72792376eca7.png)

</details>

Hopefully, this provides more insight on Android's tricky behavior.

**Comment by Pujan92** [comment]
@Victor-Nyagudi are you able to reproduce same in the snack? I am not facing this there, so doubt of bug is with forked RN repo.

**Comment by Victor-Nyagudi** [comment]
> @Victor-Nyagudi are you able to reproduce same in the snack? I am not facing this there, so doubt of bug is with forked RN repo.

Take a look at this [Snack I made](https://snack.expo.dev/ONAAh_nIY) on Android and compare it to iOS and Web. 

The text in the nested `View`  is misaligned. 

Do you think it could be some default `View` behavior I'm not aware of?

**Comment by kaushiktd** [comment]
### Please re-state the problem that we are trying to solve in this issue.
Android - Copy icon is missing in Profile/Contact method


### What is the root cause of that problem?
In React native CLI, android you can not render `<View>` component inside the `<Text>` components, this behavior is unique to Android and is caused by nesting a View inside a Text component, 
You can check inside `CopyTextToClipboard.js` file, The Tooltip component has the condition to render Icon inside the Text components, but Icon has wrapped inside View components.
Check below file:
https://github.com/Expensify/App/blob/c144663fcad41a69344311a27dd6a147a28e7191/src/components/Icon/index.js#L54-L65

not working code like below:
```
<Text>
    Hello, there 
    <View>
        <Text>I'm nested in a View</Text>
    </View>
</Text>

```

### What changes do you think we should make in order to solve the problem?
There is only one solution I tested and seems appropriate. You need to render `<View>` component outside from `<Text>` Components otherwise Copy Icon is not rendered in android.

"Add more ways for people to find you, and forward receipts to [abc@gmail.com](mailto:abc@gmail.com) + `<copy icon>`"

**Comment by MelvinBot** [comment]
@NikkiWines, @lschurr, @aimane-chnaif Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by lschurr** [comment]
@aimane-chnaif can you review the recent proposal?

**Comment by lschurr** [comment]
Still no acceptable proposals, right @aimane-chnaif?

**Comment by bernhardoj** [comment]
This has the same root cause with this https://github.com/Expensify/App/issues/17509

**Comment by bernhardoj** [comment]
Both issue comes from this PR https://github.com/Expensify/react-native/pull/37. From my testing, applying this [solution](https://github.com/Expensify/App/issues/17509#issuecomment-1525753477) will fix both issues.

**Comment by MelvinBot** [comment]
@NikkiWines @lschurr @aimane-chnaif this issue is now 3 weeks old. There is one more week left before this issue <a href='https://stackoverflowteams.com/c/expensify/questions/15233/15234#15234'>breaks WAQ</a> and will need to go internal. What needs to happen to get a PR in review this week? Please create a thread in #expensify-open-source to discuss. Thanks!

**Comment by lschurr** [comment]
Ideas for this one @NikkiWines @aimane-chnaif?

**Comment by NikkiWines** [comment]
@lschurr still looking for valid proposals I believe, going to update the price on this one!

**Comment by s-alves10** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Copy icon is missing in profile/contact page and workspace/reimburse expenses page.

### What is the root cause of that problem?

This issue happens only on Android. Below is the code segment related to copy-clipboard icon.

https://github.com/Expensify/App/blob/d120f6a9aa13dc842e265202996f8fe32019faca/src/pages/settings/Profile/Contacts/ContactMethodsPage.js#L122-L131

I've tested the same code in a new react-native project(with the latest version of react-native) and that works fine.

![image](https://user-images.githubusercontent.com/126839717/236631604-edd573f5-a485-426d-8365-0e1ff03865f3.png)

And I've found that svg icons inside the `Text` element isn't visible in the current version of react-native through a lot of tests.
```
<View>
    <CopyIcon width={16} height={16} />
</View>
```

```
<View>
    <Text>Sample text</Text>
    <CopyIcon width={16} height={16} />
</View>
```
Above code works fine but below code doesn't work.
```
<Text>
    <CopyIcon width={16} height={16} />
</Text>
```

Through these considerations, I verified that the root cause of this issue is the bug of the current version of react-native.

### What changes do you think we should make in order to solve the problem?

A tricky solution exists for this bug. Throughout the tests, I was able to find that `numberOfLines` props of `Text` component does the tricks. We need to set this value large enough to show all contents and this props should be set at the top-most `Text` component of the component tree.

Below are the code lines related.

https://github.com/Expensify/App/blob/d120f6a9aa13dc842e265202996f8fe32019faca/src/pages/settings/Profile/Contacts/ContactMethodsPage.js#L123-L130

https://github.com/Expensify/App/blob/d120f6a9aa13dc842e265202996f8fe32019faca/src/pages/workspace/reimburse/WorkspaceReimburseView.js#L237-L244

Replacing the `<Text>` at the first line with `<Text numberOfLines={5}>` fixed the issue successfully.


<details>
    <summary>Result</summary>
    

https://user-images.githubusercontent.com/126839717/236634192-81cccc82-3e96-48a5-9523-0509bdbc5ee1.mp4


</details>

### What alternative solutions did you explore? (Optional)
NA

**Comment by lschurr** [comment]
Could you review this new proposal @aimane-chnaif?

**Comment by lschurr** [comment]
How are we doing on this one @aimane-chnaif?

**Comment by NikkiWines** [comment]
Nice, @s-alves10 - thanks for the thorough investigation. 

If this is a bug in RN, we should also probably try and get that fixed in the main repo. So once we decide on a solution let's submit a change to our fork and also create an issue in the RN repo. 

However, rhe fix of modifying the top level `<Text>` to have`numberOfLInes={5}` seems a little hacky to me.

**Comment by s-alves10** [comment]
> However, the fix of modifying the top level `<Text>` to have`numberOfLInes={5}` seems a little hacky to me.

Yeah. It's a little bit tricky. We can track the bug from this solution in the main RN repo, I think.

**Comment by kbecciv** [comment]
Hello @NikkiWines! QA team is noticed that Copy icon is missing on Workspace Setting> any workspace > Reimburse expenses. Please let me know if we need to create a new ticket for this issue. 


https://github.com/Expensify/App/assets/93399543/29008cf1-0ab3-45e5-b72d-7e94cfb14de0

**Comment by s-alves10** [comment]
> still no proposals. we should wait for upstream fix

My [proposal](https://github.com/Expensify/App/issues/17368#issuecomment-1537169918) doesn't work?

**Comment by lschurr** [comment]
I'm heading OOO from May 17-23rd. I've re-added the Bug label to reassign the issue and kept myself assigned as well. If action is still needed on May 24th, I'll take it back from you @anmurali.

Latest update: still no accepted proposals.

**Comment by NikkiWines** [comment]
Still waiting on proposals here

**Comment by kaushiktd** [comment]
> > @aimane-chnaif can you review the recent proposal?
> 
> That's still not acceptable.

You can't use the View component inside Text component. This is not acceptable form native android.
https://github.com/facebook/react-native/issues/25197

So if you can state the reason for not accepting my proposal, will help me improve my technical knowledge for the same, please! 🙂 And perhaps I can make new proposal based on that.

**Comment by lschurr** [comment]
Un-assigning @anmurali since I'm back from OOO.

**Comment by lschurr** [comment]
Still waiting on proposals for this one @NikkiWines @aimane-chnaif?

**Comment by s-alves10** [comment]
> 
> Not acceptable. There's upstream [PR](https://github.com/facebook/react-native/pull/35703) but not sure how much longer we should wait.

@NikkiWines @lschurr @aimane-chnaif 
May I ask you a question?
If we have to wait [this PR](https://github.com/facebook/react-native/pull/35703) for this issue, I'm not sure why you are waiting for proposals.

I think we need to have some get-around until the [PR](https://github.com/facebook/react-native/pull/35703) is merged.
My solution [here](https://github.com/Expensify/App/issues/17368#issuecomment-1537169918) works perfectly for both contact page and reimburse page though it's a little bit tricky.

**Comment by lschurr** [comment]
Thoughts on that last comment @NikkiWines @aimane-chnaif?

**Comment by NikkiWines** [comment]
@aimane-chnaif that's true, we'd be beholden to a PR that we have no influence over. I'm hopeful that it will get merged since it's pretty active at the moment - but it's impossible to know how long that'll take. 

Given that, we should probably go with option 2 that you listed above, where we update our own fork and put in a (hopefully) temporary fix until it's fixed in the RN repo. 

However, whatever fix we put in place should still have a high standard and avoid hacky practices. Unfortunately, that leaves us back where we started 😅

 @s-alves1, since your proposal involves hardcoding `numberOfLines` to a specific value, have you tested it across various platforms with different viewport sizes to confirm that it fully resolves the issue and doesn't cause any regressions?

**Comment by s-alves10** [comment]
> @s-alves1, since your proposal involves hardcoding numberOfLines to a specific value, have you tested it across various platforms with different viewport sizes to confirm that it fully resolves the issue and doesn't cause any regressions?

Yes. I've tested both pages(contact and reimburse)

**Comment by JSExpert1121** [comment]
@NikkiWines 
Okay. Then I'll tested on various platforms and upload the video

**Comment by s-alves10** [comment]
@NikkiWines 
I'll test on various platforms and upload the video

**Comment by lschurr** [comment]
Any updates @NikkiWines @aimane-chnaif?

**Comment by s-alves10** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/17368#issuecomment-1537169918)


@aimane-chnaif 

I uploaded the demo videos. Tests were done on smaller devices(Android 480 x 320) as well

**Comment by s-alves10** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/17368#issuecomment-1537169918)


@aimane-chnaif 
I updated the proposal for social icons. And updated the videos for android native and chrome(320) reflecting the social icons

**Comment by s-alves10** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/17368#issuecomment-1537169918)

@aimane-chnaif 
I updated the proposal according to your feedback. But it has a lot of changes. Please review and let me know what you think is the best choice.

**Comment by s-alves10** [comment]
@aimane-chnaif 
Yeah, that makes sense.
I thought creating a `PressableLink` component for future use would be better.

**Comment by s-alves10** [comment]
@s-alves10 
Yes. I've checked them on 1080p/320p devices. I think you checked them already

**Comment by lschurr** [comment]
Are we assigning to @s-alves10 or what's the next step?

**Comment by NikkiWines** [comment]
Oh sorry, missed this ping - looking now

**Comment by NikkiWines** [comment]
@s-alves10 can you confirm that this change works even if you modify the text size (i.e. zooming in and out and also manually changing the text size on the mobile device) 

If that works, then I think we can go ahead and use this solution. Sorry for the delay 🙇

**Comment by s-alves10** [comment]
@NikkiWines 

I can confirm.

`numOfLines` defines the maximum number of lines. If the text is longer or larger, ellipsis will appear. If the text is shorter, `numberOfLines` doesn't affect the component size.

So we can set the `numberOfLines` to large number, i.e. 100. This works perfect.

**Comment by NikkiWines** [comment]
Ok cool, thanks @s-alves10. In that case, let's move forward with your proposal.

**Comment by s-alves10** [comment]
@aimane-chnaif 
[PR](https://github.com/Expensify/App/pull/20528) is ready for review

**Comment by s-alves10** [comment]
@NikkiWines 
The upwork job link is no longer available.

**Comment by NikkiWines** [comment]
@lschurr can we re-post the upwork job for this please 🙇

**Comment by lschurr** [comment]
New open job posting is here: https://www.upwork.com/jobs/~01ac7e80f8af912eb4

**Comment by s-alves10** [comment]
I submitted a proposal. Thank you

**Comment by lschurr** [comment]
@aimane-chnaif - please apply for the job here: https://www.upwork.com/jobs/~01ac7e80f8af912eb4

Do we need a regression test for this?

**Comment by lschurr** [comment]
Payment sent to @s-alves10.

**Comment by s-alves10** [comment]
Thank you

**Comment by lschurr** [comment]
Be sure to apply for the job @aimane-chnaif - https://www.upwork.com/jobs/~01ac7e80f8af912eb4

**Comment by lschurr** [comment]
Paid with bonuses. Closing!
